lsp/nls/src/position.rs

```diff
@@ -0,0 +1,162 @@
+use codespan::ByteIndex;
+use nickel_lang_core::{
+    position::TermPos,
+    term::{RichTerm, SharedTerm, Traverse, TraverseControl},
+};
+use range_map::{Range, RangeMap};
+
+// A term that uses pointer equality to implement Eq.
+#[derive(Clone, Debug)]
+struct RichTermPtr(RichTerm);
+
+impl PartialEq for RichTermPtr {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.pos == other.0.pos && SharedTerm::ptr_eq(&self.0.term, &other.0.term)
+    }
+}
+
+impl Eq for RichTermPtr {}
+
+/// A lookup data structure, for looking up the term at a given position.
+///
+/// Overlapping positions are resolved in favor of the smaller one; i.e., lookups return the
+/// most specific term for a given position.
+#[derive(Clone, Debug)]
+pub struct PositionLookup {
+    table: RangeMap<u32, RichTermPtr>,
+}
+
+impl PositionLookup {
+    /// Create a position lookup table for looking up subterms of `rt` based on their positions.
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut all_ranges = Vec::new();
+
+        rt.traverse_ref(&mut |term: &RichTerm| {
+            if let TermPos::Original(pos) = &term.pos {
+                all_ranges.push((pos.start.0, pos.end.0, term.clone()));
+            }
+            TraverseControl::<()>::Continue
+        });
+
+        // We rely on the invariant that if two ranges overlap then one is contained in the
+        // other. That is, the ranges form a tree. This sort order corresponds to pre-order
+        // traversal of that tree.
+        //
+        // (It would be nice if we could build the lookup table directly from the traversal
+        // in traverse_ref, but (1) the state tracking is fiddly and (2) the traversal order
+        // in traverse_ref can be weird; for example, a `Term::Annotated` sometimes contains
+        // children whose position is outside the position of the parent.)
+        all_ranges.sort_by_key(|range| (range.0, std::cmp::Reverse(range.1)));
+        let mut all_ranges = all_ranges.into_iter().peekable();
+
+        // Now we iterate over the tree of ranges to make a disjoint set of ranges.
+        // `stack` is the path in the tree leading to the node that we are currently visiting.
+        let mut stack = Vec::new();
+        let mut disjoint = Vec::new();
+        let mut maybe_cur = all_ranges.next();
+        let mut pos = maybe_cur.as_ref().map(|range| range.0).unwrap_or_default();
+
+        // We accumulate ranges using this closure, which guarantees that they are non-overlapping.
+        // (This is important: RangeMap panics if they are not.)
+        let mut push_range = |end: u32, term: RichTerm| {
+            if pos < end {
```

Comment:
It does happen that `pos == end`, definitely. For example, if a parent has range [0, 10) and its children have adjacent ranges [1, 5) and [5, 8) then after visiting the first child we will have `pos == 5`. We'll revisit the parent (because we don't know that the children are adjacent; there could be a gap where the parent is visible again) and try to push an empty range.

However, I agree it shouldn't happen that `pos > end`, so I'll add a debug_assert.

Reply:
I think I'll give a second look at the code to understand your answer, but right now it's not totally clear to me why we would be visiting the parent again. I thought we were just 1) adding all intervals from the rich term in a datastructure (so one interval per AST node) 2) sort that and post-process it in some way, visiting each interval once.